source/common/formatter/substitution_formatter.h

```diff
@@ -241,6 +243,95 @@ class StructFormatter {
 
 using StructFormatterPtr = std::unique_ptr<StructFormatter>;
 
+// Helper classes for OpenTelemetryFormatter::OpenTelemetryFormatMapVisitor.
+template <class... Ts> struct OpenTelemetryFormatMapVisitorHelper : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+OpenTelemetryFormatMapVisitorHelper(Ts...) -> OpenTelemetryFormatMapVisitorHelper<Ts...>;
+
+/**
+ * An formatter for structured log formats, which returns a Struct proto that
+ * can be converted easily into multiple formats.
+ */
+class OpenTelemetryFormatter {
+public:
+  OpenTelemetryFormatter(const ::opentelemetry::proto::common::v1::KeyValueList& format_mapping,
+                         bool omit_empty_values);
+  OpenTelemetryFormatter(const ::opentelemetry::proto::common::v1::KeyValueList& format_mapping,
+                         bool omit_empty_values, const std::vector<CommandParserPtr>& commands);
+
+  ::opentelemetry::proto::common::v1::KeyValueList
+  format(const Http::RequestHeaderMap& request_headers,
+         const Http::ResponseHeaderMap& response_headers,
+         const Http::ResponseTrailerMap& response_trailers,
+         const StreamInfo::StreamInfo& stream_info, absl::string_view local_reply_body) const;
+
+private:
+  struct OpenTelemetryFormatMapWrapper;
+  struct OpenTelemetryFormatListWrapper;
+  using OpenTelemetryFormatValue =
+      absl::variant<const std::vector<FormatterProviderPtr>, const OpenTelemetryFormatMapWrapper,
+                    const OpenTelemetryFormatListWrapper>;
+  // Although not required for Struct/JSON, it is nice to have the order of
+  // properties preserved between the format and the log entry, thus std::map.
+  using OpenTelemetryFormatMap = std::map<std::string, OpenTelemetryFormatValue>;
+  using OpenTelemetryFormatMapPtr = std::unique_ptr<OpenTelemetryFormatMap>;
+  struct OpenTelemetryFormatMapWrapper {
+    OpenTelemetryFormatMapPtr value_;
+  };
```

Comment:
A drive-be question: what's the benefit of `OpenTelemetryFormatMapWrapper` over `OpenTelemetryFormatMapPtr`?

Reply:
I just copied the whole StructFormatter as is and converted the types to match OpenTelemetry logging.
I tried templating it but it adds a lot of boiler plate which I'm not sure is worth the trouble.